lib/core/application/chat/new_chat_bloc/new_chat_bloc.dart

```diff
@@ -0,0 +1,119 @@
+import 'package:app/core/presentation/widgets/future_loading_dialog.dart';
+import 'package:app/core/service/matrix/matrix_service.dart';
+import 'package:app/injection/register_module.dart';
+import 'package:auto_route/auto_route.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter_bloc/flutter_bloc.dart';
+import 'package:freezed_annotation/freezed_annotation.dart';
+import 'package:matrix/matrix.dart';
+import 'package:matrix/matrix.dart' as sdk;
+
+part 'new_chat_bloc.freezed.dart';
+
+class NewChatBloc extends Bloc<NewChatEvent, NewChatState> {
+  final matrixService = getIt<MatrixService>();
+
+  NewChatBloc()
+      : super(
+          const NewChatState(
+            userSearchResult: null,
+            selectedUsers: [],
+            isSearching: false,
+          ),
+        ) {
+    on<NewChatEventSearchUsers>(_onSearchUsers);
+    on<NewChatEventSelectUser>(_onSelectUser);
+    on<NewChatEventDeselectUser>(_onDeselectUser);
+  }
+
+  _onSearchUsers(NewChatEventSearchUsers event, Emitter emit) async {
+    emit(state.copyWith(isSearching: true));
+    var text = event.text;
+    try {
+      final result = await matrixService.client.searchUserDirectory(
+        text!,
+        limit: 20,
+      );
+      emit(state.copyWith(isSearching: false));
+      emit(state.copyWith(userSearchResult: result));
+    } catch (e) {
+      emit(state.copyWith(isSearching: false));
+    }
+  }
+
+  _onSelectUser(NewChatEventSelectUser event, Emitter emit) {
+    final selectedUsers = List<Profile>.from(state.selectedUsers);
+    selectedUsers.add(event.user);
+    emit(state.copyWith(selectedUsers: selectedUsers));
+  }
+
+  _onDeselectUser(NewChatEventDeselectUser event, Emitter emit) {
+    final selectedUsers = List<Profile>.from(state.selectedUsers);
+    selectedUsers.remove(event.user);
+    emit(state.copyWith(selectedUsers: selectedUsers));
+  }
+
+  startChat(BuildContext context) async {
+    final List<String> selectedUserIds =
+        state.selectedUsers.map((profile) => profile.userId).toList();
```

Comment:
Avoid using `context` inside BloC class, the purpose of the BloC is to separate logic and UI. This is violated the principles of BloC
Try emit an event back to UI and handle navigation at there using BlocListener

Reply:
Thanks 👍